From 82883b7fa26e0120f95f8d0f638f122e8b3e70b4 Mon Sep 17 00:00:00 2001
From: Ravi Babu <ravibabu@ti.com>
Date: Wed, 29 May 2013 18:37:04 +0530
Subject: [PATCH 2/3] usb: musb: core: added babble recovery func-ptr to
 musb->ops

Adding babble recovery mechanism for dsps platform and as part
of babble recovery the controller will be restarted.

Signed-off-by: Ravi Babu <ravibabu@ti.com>
---
 drivers/usb/musb/musb_core.c | 6 ++++++
 drivers/usb/musb/musb_core.h | 7 +++++++
 2 files changed, 13 insertions(+)

diff --git a/drivers/usb/musb/musb_core.c b/drivers/usb/musb/musb_core.c
index 4b899dd..bf848ab 100644
--- a/drivers/usb/musb/musb_core.c
+++ b/drivers/usb/musb/musb_core.c
@@ -848,6 +848,12 @@ b_host:
 		}
 	}
 
+	/* handle babble condition */
+	if (int_usb & MUSB_INTR_BABBLE) {
+		pr_info("babble: restarting the musb controller..");
+		musb_babble_recovery(musb);
+	}
+
 #if 0
 /* REVISIT ... this would be for multiplexing periodic endpoints, or
  * supporting transfer phasing to prevent exceeding ISO bandwidth
diff --git a/drivers/usb/musb/musb_core.h b/drivers/usb/musb/musb_core.h
index 6d9262e..8b88326 100644
--- a/drivers/usb/musb/musb_core.h
+++ b/drivers/usb/musb/musb_core.h
@@ -199,6 +199,8 @@ struct musb_platform_ops {
 	int	(*adjust_channel_params)(struct dma_channel *channel,
 				u16 packet_sz, u8 *mode,
 				dma_addr_t *dma_addr, u32 *len);
+
+	void	(*babble_recovery)(struct musb *musb);
 };
 
 /*
@@ -577,4 +579,9 @@ static inline int musb_platform_exit(struct musb *musb)
 	return musb->ops->exit(musb);
 }
 
+static inline void musb_babble_recovery(struct musb *musb)
+{
+	if (musb->ops->babble_recovery)
+		musb->ops->babble_recovery(musb);
+}
 #endif	/* __MUSB_CORE_H__ */
-- 
2.1.0.rc1

